File: app/api/v1/routes/tokens.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""Scoped read-only API tokens (§2.4) — PIN-gated management.

Create/revoke require an authenticated session (require_session), so a read-only API token can
neither mint nor revoke tokens (it is rejected with 403 on those mutating calls). The raw
token is returned once at creation and is never retrievable afterwards.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_db, require_session
from app.models import AuditEvent
from app.services.api_tokens import create_token, list_tokens, revoke_token

router = APIRouter()


class TokenIn(BaseModel):
    name: str = Field(default="API token", max_length=80)


@router.get("/tokens", dependencies=[Depends(require_session)])
async def get_tokens(session: AsyncSession = Depends(get_db)) -> dict:
    """List token metadata (never the raw token or its hash)."""
    return {"tokens": await list_tokens(session)}


@router.post("/tokens", dependencies=[Depends(require_session)])
async def create(payload: TokenIn, session: AsyncSession = Depends(get_db)) -> dict:
    tok, raw = await create_token(session, payload.name)
    session.add(AuditEvent(category="security", action="api_token_created", detail=tok.prefix))
    await session.commit()
    return {"id": tok.id, "name": tok.name, "prefix": tok.prefix, "token": raw,
            "note": "Store this now — it is shown only once and cannot be retrieved. "
                    "Use it read-only: Authorization: Token <token> on GET requests."}


@router.delete("/tokens/{token_id}", dependencies=[Depends(require_session)])
async def revoke(token_id: int, session: AsyncSession = Depends(get_db)) -> dict:
    if not await revoke_token(session, token_id):
        raise HTTPException(404, "token not found")
    session.add(AuditEvent(category="security", action="api_token_revoked", detail=str(token_id)))
    await session.commit()
    return {"ok": True}

File: frontend/src/api/tokens.ts
// API-token card (page-settings Privacy tab; D-069 / §9-8). Create/list/revoke go through
// /tokens, which is require_session — a read-only API token can neither mint nor revoke tokens
// (403). The raw token is returned ONCE at creation and never retrievable after (tokens.py).
import { apiGet, apiSend } from "./client";

export interface TokenMeta {
  id: number;
  name: string;
  prefix: string;
  /** ISO timestamp or null; rendered as a served string (D-105). A never-used token → null → em dash. */
  created_at: string | null;
  last_used_at: string | null;
}

export interface CreatedToken {
  id: number;
  name: string;
  prefix: string;
  /** The raw token — shown exactly once at creation; never re-read. */
  token: string;
  note: string;
}

export async function listTokens(): Promise<TokenMeta[] | null> {
  const r = await apiGet<{ tokens: TokenMeta[] }>("/tokens");
  return r.ok ? r.data.tokens : null;
}

export async function createToken(
  name: string,
): Promise<{ ok: true; token: CreatedToken } | { ok: false; error: string }> {
  const r = await apiSend<CreatedToken>("/tokens", "POST", { name });
  return r.ok ? { ok: true, token: r.data } : { ok: false, error: r.error };
}

/** Revoke a token — require_session (§9-8): the SESSION, not a fresh PIN. D-103 binds only the
 *  destructive purge; a revoked token is re-creatable, so revoke does not take a fresh PIN. */
export async function revokeToken(
  id: number,
): Promise<{ ok: true } | { ok: false; error: string }> {
  const r = await apiSend<{ ok?: boolean }>(`/tokens/${id}`, "DELETE");
  return r.ok ? { ok: true } : { ok: false, error: r.error };
}
